feat(generic): Support setting the Electron app path in start()

bendemboski opened the pull request:
<!--
Thanks for filing a pull request!
Please check off all of the steps as they are completed by replacing [ ] with [x].
-->

* [x] I have read the [contribution documentation](https://github.com/electron-userland/electron-forge/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [code of conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project follows, as appropriate.
* [x] The changes are appropriately documented (if applicable).
* [x] The changes have sufficient test coverage (if applicable).
* [x] The testsuite passes successfully on my local machine (if applicable).

**Summarize your changes:**

The start command allows the caller to specify the directory for the Electron forge project, and then assumes that its main module is the Electron application. However, some projects may contain multiple Electron applications, or may want to specify an html file or a URL as the Electron application (see [here](https://github.com/electron/electron/blob/cf694ef32b78f0904219c6c8ba554d3d5cbea037/default_app/main.js#L338)).

So, this adds an optional 'appPath' API option (and corresponding '-p'/'--app-path' CLI option) that is the path to the Electron application relative to the electron forge project (the 'dir' option).

This also fixes the start command-line options parsing, which was passing all arguments through to the Electron application instead of parsing out the start command options.

NOTE: This allows ember-electron to use electron-forge to implement its test command, and is part of that migration effort.

MarshallOfSound commented on src/electron-forge-start.js:
```diff
@@ -25,6 +26,7 @@ import { start } from './api';
     dir,
     interactive: true,
     enableLogging: program.enableLogging,
+    appPath: program.appPath,
```
Passing in `undefined` here will break the default because

```js
Object.assign({}, { a: 1 }, { a: undefined }).a === undefined
```

bendemboski replied:
Ah, yes, you are right. Sorry, @MarshallOfSound, sloppy of me.

I did some more detailed testing, and it looks like the start command's command parsing is pretty broken. None of the command-line options will ever get parsed because [this](https://github.com/electron-userland/electron-forge/blob/master/src/electron-forge-start.js#L22) line slices it down to `[ '<path>/node', '<path>/electron-forge-start' ]`. So any command-line arguments meant for the start command itself will not be handled, and instead will get [passed through](https://github.com/electron-userland/electron-forge/blob/master/src/electron-forge-start.js#L28) to the electron application itself.

It looks like the fix would be to change the `dir` parameter to be an option (`-d dir`) so we can tell it apart from the `args` to be passed to the electron app. I'd be happy to fix this in my PR and get some unit tests in place, but being a new contributor to this project, I'd like a nod from somebody already involved first.

MarshallOfSound replied:
Good catch, I wonder if using `allowUnknownOption` somehow reveals any unparsed options?

malept commented on src/electron-forge-start.js:
```diff
@@ -10,7 +10,9 @@ import { start } from './api';
   program
     .version(require('../package.json').version)
     .arguments('[cwd]')
+    .option('-p, --app-path <path>', 'Override the path to the electron app to launch (defaults to \'.\')')
```
Does our linter setup complain if the description string here is double-quoted? I want to avoid escaped quotes for readability.

malept commented on src/electron-forge-start.js:
```diff
@@ -10,7 +10,9 @@ import { start } from './api';
     .arguments('[cwd]')
+    .option('-p, --app-path <path>', 'Override the path to the electron app to launch (defaults to \'.\')')
     .option('-l, --enable-logging', 'Enable advanced logging.  This will log internal Electron things')
+    .option('-a, --app-args <args>', 'Command-line arguments to pass to the electron application (surround multiple with quotes)')
```
Capitalize Electron, please.

malept commented on src/electron-forge-start.js:
```diff
@@ -19,12 +21,22 @@ import { start } from './api';
+    opts.enableLogging = program.enableLogging;
+  }
+  if (program.appArgs) {
+    opts.args = program.appArgs.split(/\s+/);
```
Really hope there aren't quoted app args here 😵

malept commented on test/fast/start_spec.js:
```diff
@@ -35,6 +35,27 @@ describe('start', () => {
     expect(spawnStub.firstCall.args[2].env).to.not.have.property('ELECTRON_ENABLE_LOGGING');
   });
 
+  it('should pass electron \'.\' as the app path if not specified', async () => {
```
Same quoting question here.

malept commented on test/fast/electron_forge_start_spec.js:
```diff
@@ -0,0 +1,97 @@
+    });
+  });
+
+  it('should handle a multiple app args', async () => {
```
```javascript
it('should handle multiple app args', [...]
```

MarshallOfSound commented:
This now isn't as fluid as I would like, having to do `--app-args` and then quote everything correctly seems like it would cause a bit of trouble.  Perhaps we can manually strip out the from the argv the params that the `start` command handles.

bendemboski commented:
@malept @MarshallOfSound I believe I've addressed your feedback. If you have any more, let me know. If it looks good, then please give me a heads-up before merging as I'd like the chance to rebase some of the code review feedback commits for a cleaner git history before it hits master.

MarshallOfSound commented:
@bendemboski Looks good to me, just need to squash and pull in latest master 👍
